File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
#bfs: 깊이우선탐색
#덱을 활용, ->큐생성하고
# # 기준값을 기준으로 갈 수 있는 방향이 있으면 추가하고, pop하고 그 인접값 탐색해서 또 넣고.....
from collections import deque

def bfs(mapp,x,y):
    queue=deque([(x,y)]) #큐에 데이터 넣기(x,y)로 묶인걸 []안에 넣어야함!!!
    dx=[-1,0,1,0]
    dy=[0,-1,0,1]

    while queue:
        (x,y)=queue.popleft() #튜플형태로 묶어줘야함!!
        for k in range(4):
            nx=x+dx[k]
            ny=y+dy[k]
            if 0<=nx<n and 0<=ny<m and mapp[nx][ny]==1:
                mapp[nx][ny]=mapp[x][y]+1 #이동하는 칸수 더하기
                queue.append([nx,ny])
    return mapp[n-1][m-1]

                
n,m=map(int,input().split()) #행,열
mapp=[list(map(int,input())) for i in range(n)]

a=bfs(mapp,0,0)
print(a)

File: 백준/Silver/2178. 미로 탐색/미로 탐색.java


import java.io.*;
import java.util.*;

public class Main {
	static int [][] arr;
	static int [][] visited;
	static int n,m;
	
	
	
	public static void bfs(int x, int y) {
		Queue<int[]> q = new LinkedList<>();
		int[] dx = new int[] {-1,0,1,0};
		int[] dy = new int[] {0,-1,0,1};
		q.add(new int[] {x,y}); //큐에 배열을 넣는 방법->q의 타입을 배열로 정의하고, 객체만들 때 x,y를 넣어주기

		visited[x][y]=1;
		
		while(!q.isEmpty()) {
			int[] temp = q.poll();
			int i = temp[0];
			int j =temp[1];
//			if (i==n-1 && j==m-1) {
//				break;
//			}
			
			for (int k=0; k<4; k++) {
				int nx = i+dx[k];
				int ny = j+dy[k];
				if (nx>=0 && nx<n && ny>=0 && ny<m) { //방문안했고, 1(길이 있으면)
					if (visited[nx][ny]==0 && arr[nx][ny]==1) {
						visited[nx][ny]=visited[i][j]+1;
						q.add(new int[] {nx,ny});
					}
				}
			}
		}
	}
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String nm = br.readLine();
		n =Integer.parseInt(nm.split(" ")[0]);
		m =Integer.parseInt(nm.split(" ")[1]);
		arr= new int [n][m];
		visited= new int[n][m]; 
		for (int i=0; i<n; i++) {
			String temp=br.readLine();
			for (int j=0; j<m; j++) {
				arr[i][j]=temp.charAt(j)-'0';
			}
		}
		bfs(0,0);
		System.out.println(visited[n-1][m-1]);
	}
}



	